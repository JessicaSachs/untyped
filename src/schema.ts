import { getType, isObject } from './utils'
import type { InputObject, InputValue, JSValue, Schema } from './types'

export function resolveSchema (obj: InputObject) {
  const schema = _resolveSchema(obj, obj, obj)
  // TODO: Create meta-schema fror superset of Schema interface
  // schema.$schema = 'http://json-schema.org/schema#'
  return schema
}

function _resolveSchema (input: InputValue, parent: InputObject, root: InputObject): Schema {
  // Node is plain value
  if (!isObject(input)) {
    return {
      type: getType(input),
      default: input as JSValue
    }
  }

  // Clone to avoid mutation
  const node = { ...input as any } as InputObject
  const schema: Schema = { ...node.$schema }

  // Resolve children
  for (const key in node) {
    // Ignore special keys
    if (key === '$resolve' || key === '$schema' || key === '$default') {
      continue
    }
    const resolved = _resolveSchema(node[key], node, root)
    schema.properties = schema.properties || {}
    schema.properties[key] = resolved
    node[key] = resolved.default
  }

  // Infer default value from $resolve and $default
  if ('$default' in node) {
    schema.default = node.$default
  }
  if (typeof node.$resolve === 'function') {
    schema.default = node.$resolve(schema.default, parent, root)
  }

  // Infer type from default value
  if (!schema.type) {
    if (Array.isArray(schema.default)) {
      schema.type = 'array'
      schema.items = schema.default.map(i => getType(i)).map(type => ({ type }))
    } else {
      schema.type = getType(schema.default) || (schema.properties ? 'object' : 'any')
    }
  }

  return schema
}
