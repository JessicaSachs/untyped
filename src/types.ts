export type JSValue =
  string |
  number |
  bigint |
  boolean |
  symbol |
  Function |
  Array<any> |
  undefined |
  null

export type JSType =
  'string' |
  'number' |
  'bigint' |
  'boolean' |
  'symbol' |
  'function' |
  'object' |
  'any' |
  'array'

// A subset of JSONSchema7
export interface Schema {
  type?: JSType | JSType[]
  items?: Schema
  default?: JSValue
  properties?: { [key: string]: Schema }
  title?: string
  description?: string
  $schema?: string
}

export interface InputObject {
  [key: string]: any
  $schema?: Schema
  $resolve?: JSValue | ((value: any, parent: InputObject, root: InputObject) => JSValue)
}

export type InputValue = InputObject | JSValue
