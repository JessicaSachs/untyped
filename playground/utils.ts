import { reactive, watch, computed } from 'vue'

export function evaluateSource (src) {
  let val
  // eslint-disable-next-line no-eval
  eval('val = ' + src.replace('export default', ''))
  return val
}

export function tryFn (fn) {
  try {
    return fn()
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error(err)
    return null
  }
}

export function persistedState (initialState = {}) {
  const state = reactive({
    ...initialState,
    ...tryFn(() => JSON.parse(atob(window.location.hash.substr(1))))
  })
  watch(state, () => {
    window.location.hash = '#' + btoa(JSON.stringify(state))
  })
  return state
}

export function safeComputed (fn) {
  let lastState = null
  return computed(() => {
    try {
      lastState = fn()
      return lastState
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error(err)
      return lastState
    }
  })
}

export const defaultInput = `export default {
  name: 'earth',
  specs: {
    gravity: {
      $resolve: val => parseFloat(val),
      $default: '9.8'
    },
    moons: {
      $resolve: (val = ['moon']) => [].concat(val),
      $schema: {
        title: 'planet moons'
      }
    }
  }
}`
