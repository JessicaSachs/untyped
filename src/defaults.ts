import { resolveSchema } from './schema'
import type { InputObject, Schema } from './types'

export function applyDefaults (ref: InputObject, input: InputObject) {
  const schema = resolveSchema(ref, true)
  return _applyDefaults(schema, input)
}

export function _applyDefaults (schema: Schema, input: InputObject, root?: InputObject) {
  if (schema.type !== 'object') {
    return {}
  }
  const merged = {}
  for (const key in schema.properties!) {
    const prop = schema.properties[key]
    if (prop.type === 'object') {
      merged[key] = _applyDefaults(prop, input?.[key], root || input)
    } else if (typeof prop.resolve === 'function') {
      merged[key] = prop.resolve(input?.[key], merged, root || input)
    } else {
      merged[key] = input?.[key] || prop.default
    }
  }
  return merged
}
