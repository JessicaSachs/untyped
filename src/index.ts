export { resolveSchema } from './schema'
export { generateDts } from './dts'

export * from './types'
