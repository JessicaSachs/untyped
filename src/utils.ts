import type { JSType } from './types'

export function escapeKey (val: string): string {
  return /^\w+$/.test(val) ? val : `"${val}"`
}

export function getType (val: any): JSType | null {
  const type = typeof val
  if (type === 'undefined' || val === null) {
    return null
  }
  if (Array.isArray(val)) {
    return 'array'
  }
  return type
}

export function isObject (val: any): boolean {
  return val !== null && !Array.isArray(val) && typeof val === 'object'
}

export function unique (arr: any[]) {
  return Array.from(new Set(arr))
}
