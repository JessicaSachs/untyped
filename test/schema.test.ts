// import { posix as p } from 'path'
import { resolveSchema, generateDts } from '../src'
import type { InputObject } from '../src/types'
import { expectSnapshot } from './utils'

const fixture1: InputObject = {
  name: 'earth',
  specs: {
    gravity: {
      $resolve: val => parseFloat(val),
      $default: '9.8'
    },
    moons: {
      $resolve: (val = ['moon']) => [].concat(val),
      $schema: {
        title: 'planet moons'
      }
    }
  }
}

// const fixture2: InputObject = {
//   srcDir: 'src',
//   userOptions: {
//     $schema: {
//       title: 'custom user options'
//     }
//   },
//   workingDir: {
//     $default: '/',
//     $schema: {
//       title: 'working directory'
//     }
//   },
//   entry: {
//     $resolve: (val: string = 'index.js', parent: any) => p.resolve(parent.workingDir, parent.srcDir, val),
//     $schema: {
//       title: 'entrty file',
//       description: 'will be resolved from current working directory if not absolute'
//     }
//   }
// }

describe('schema', () => {
  it('resolveSchema', () => {
    const schema = resolveSchema(fixture1)
    expectSnapshot(schema, 'schema.json')
  })
})

describe('dts', () => {
  it('generateDts', () => {
    const schema = resolveSchema(fixture1)
    const dts = generateDts(schema)
    expectSnapshot(dts, 'fixture.gen.d.ts')
  })
})
