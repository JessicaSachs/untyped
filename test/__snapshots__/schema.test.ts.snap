// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dts generateDts 1`] = `
"interface MyObject {
   /** @default \\"earth\\" */
  name: string,

  specs: {
    /** @default 9.8 */
    gravity: number,

    /**
     * planet moons
     * @default [\\"moon\\"]
    */
    moons: string[],
  },
}"
`;

exports[`schema resolveSchema 1`] = `
"{
  \\"properties\\": {
    \\"name\\": {
      \\"type\\": \\"string\\",
      \\"default\\": \\"earth\\"
    },
    \\"specs\\": {
      \\"properties\\": {
        \\"gravity\\": {
          \\"default\\": 9.8,
          \\"type\\": \\"number\\"
        },
        \\"moons\\": {
          \\"title\\": \\"planet moons\\",
          \\"default\\": [
            \\"moon\\"
          ],
          \\"type\\": \\"array\\",
          \\"items\\": [
            {
              \\"type\\": \\"string\\"
            }
          ]
        }
      },
      \\"type\\": \\"object\\"
    }
  },
  \\"type\\": \\"object\\"
}"
`;
